### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

As final considerations:

* You must use the Pandas Library and the Jupyter Notebook.
* You must submit a link to your Jupyter Notebook with the viewable Data Frames.
* You must include a written description of three observable trends based on the data.
* See [Example Solution](HeroesOfPymoli/HeroesOfPymoli_starter.ipynb) for a reference on expected format.

In [31]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
data_file = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(data_file)

In [32]:
#Show first five rows of data

purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [33]:
#Count each of the total_players in SN column; .value_counts() counts unique values
total_players = len(purchase_data["SN"].value_counts())
total_players

#Convert the total_players count to DataFrame where dictionary key: Player Count and value: total_players
total_players_df = pd.DataFrame({"Player Count":[total_players]})
total_players_df

Player Count
0           576

### Purchasing Analysis (Total)

* Number of Unique Items
* Average Purchase Price
* Total Number of Purchases
* Total Revenue

* Run basic calculations to obtain number of unique items, average price, etc.
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame


In [34]:

#Calculate number of unique items ("Item ID")
number_unique_items = len(purchase_data["Item ID"].value_counts())

# #Caluculate average price ("Price")
average_price = round(purchase_data["Price"].mean(),2)

# #Calculate number of purchases ("Purchase ID")
number_purchases = len(purchase_data["Purchase ID"].value_counts())

#Calculate number of total revenue = sum of ("Price")
total_revenue = purchase_data["Price"].sum()


# #create a summary data frame to hold the results
analysis_summary = pd.DataFrame({"Number of Unique Items":[number_unique_items], "Average Price":[average_price], "Number of Purchases":[number_purchases],"Total Revenue":[total_revenue]})

#Format columns using .map("${:,.2f}".format)
analysis_summary["Average Price"] = analysis_summary["Average Price"].map("${:,.2f}".format)
analysis_summary["Total Revenue"] = analysis_summary["Total Revenue"].map("${:,.2f}".format)

# #Display the summary data frame
analysis_summary

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics

Percentage and Count of Male Players
Percentage and Count of Female Players
Percentage and Count of Other / Non-Disclosed


In [35]:
#Calculate count of all observations by column
#***This count is inaccurate; it counts players more than once in gender counts***
purchase_data.count()


Purchase ID    780
SN             780
Age            780
Gender         780
Item ID        780
Item Name      780
Price          780
dtype: int64

In [36]:
#Calculate total number of Players by Gender category
#***This count is inaccurate; it counts players more than once in gender counts***
total_gender_count = purchase_data["Gender"].value_counts()
total_gender_count


Male                     652
Female                   113
Other / Non-Disclosed     15
Name: Gender, dtype: int64

In [37]:
#Calculate unique number of male players
unique_male_players = purchase_data[purchase_data.Gender=='Male'].SN.value_counts()
unique_male_players_total = unique_male_players.count()
print(unique_male_players_total)


#Calculate percentage of male players
male_percentage = round(unique_male_players_total / total_players, 2)
print(male_percentage)

484
0.84


In [38]:
#Calculate unique number of female players
unique_female_players = purchase_data[purchase_data.Gender=='Female'].SN.value_counts()
unique_female_players_total = unique_female_players.count()
print(unique_female_players_total)

#Calculate percentage of female players
female_percentage = round(unique_female_players_total / total_players, 2)
print(female_percentage)

81
0.14


In [39]:
#Calculate unique number of Other / Non-Disclosed players
unique_other_players = purchase_data[purchase_data.Gender=='Other / Non-Disclosed'].SN.value_counts()
unique_other_players_total = unique_other_players.count()
print(unique_other_players_total)

#Calculate percentage of undiclosed gender players
other_percentage = round(unique_other_players_total / total_players, 2)
print(other_percentage)

11
0.02


In [40]:
#Confirm all gender counts equal variable, total_players: 576 (Player Count section at beginning)

unique_players_gender_total = unique_male_players_total + unique_female_players_total + unique_other_players_total
print(unique_players_gender_total)

total_percent = male_percentage + female_percentage + other_percentage
print(total_percent)

576
1.0


In [41]:
#create a summary data frame to hold the Gender Demographics results
unique_players_df = pd.DataFrame({"Unique Male Players":[unique_male_players_total, male_percentage], "Unique Female Players":[unique_female_players_total, female_percentage], "Undisclosed Gender Players":[unique_other_players_total, other_percentage], "Total":[unique_players_gender_total, total_percent],})

unique_players_df
unique_players_df

Unique Male Players  Unique Female Players  Undisclosed Gender Players  \
0               484.00                  81.00                       11.00   
1                 0.84                   0.14                        0.02   

   Total  
0  576.0  
1    1.0


## Purchasing Analysis (Gender)

Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender
Create a summary data frame to hold the results
Optional: give the displayed data cleaner formatting
Display the summary data frame

* The below each broken by gender
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Gender




In [42]:
purchase_data.columns

Index(['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price'], dtype='object')

In [43]:
#GroupBy gender
grouped_gender = purchase_data.groupby(["Gender"])

In [44]:
#Calculate purchase count by gender
g_purchase_count = grouped_gender["Purchase ID"].count()
g_purchase_count

Gender
Female                   113
Male                     652
Other / Non-Disclosed     15
Name: Purchase ID, dtype: int64

In [45]:
#Calculate average purchase price per person by gender

avg_purchase_price = round(grouped_gender.Price.mean(),2)
avg_purchase_price


Gender
Female                   3.20
Male                     3.02
Other / Non-Disclosed    3.35
Name: Price, dtype: float64

In [46]:
#Calculate average total purchase price per person by gender
total_purchase_value = round(grouped_gender.Price.sum(),2)
total_purchase_value


Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

In [47]:
purchasing_analysis_df = pd.DataFrame({"Purchase Count": g_purchase_count, "Average Purchase Price per Person by Gender": avg_purchase_price, "Total Purchase Value": total_purchase_value})


purchasing_analysis_df["Average Purchase Price per Person by Gender"] = purchasing_analysis_df["Average Purchase Price per Person by Gender"].map("${:,.2f}".format)
purchasing_analysis_df["Total Purchase Value"] = purchasing_analysis_df["Total Purchase Value"].map("${:,.2f}".format)

purchasing_analysis_df

Purchase Count  \
Gender                                  
Female                            113   
Male                              652   
Other / Non-Disclosed              15   

                      Average Purchase Price per Person by Gender  \
Gender                                                              
Female                                                      $3.20   
Male                                                        $3.02   
Other / Non-Disclosed                                       $3.35   

                      Total Purchase Value  
Gender                                      
Female                             $361.94  
Male                             $1,967.64  
Other / Non-Disclosed               $50.19

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [48]:
purchase_data.columns

Index(['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price'], dtype='object')

In [49]:
# Create age bins
age_bin = [0, 9.90, 14.90, 19.90, 24.90, 29.90, 34.90, 39.90, 99999]
age_labels = ["<10", "10-14","15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

In [50]:
#Add new column and categorize the existing players using the age bins. Hint: use pd.cut()
purchase_data["Age Ranges"] = pd.cut(purchase_data["Age"], age_bin, labels=age_labels)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Ranges  
0  Extraction, Quickblade Of Trembling Hands   3.53      20-24  
1                          Frenzied Scimitar   1.56        40+  
2                               Final Critic   4.88      20-24  
3                                Blindscythe   3.27      20-24  
4                                       Fury   1.44      20-24

In [51]:
#Calculate the numbers and percentages by age group
gen_group = purchase_data.groupby(["Age Ranges"])

#Calculate purchase count by gender
gen_group_count = gen_group["SN"].count()

#Create a summary data frame to hold the results
gen_group_count_df = pd.DataFrame(gen_group_count)

#Display the summary data frame
gen_group_count_df

SN
Age Ranges     
<10          23
10-14        28
15-19       136
20-24       365
25-29       101
30-34        73
35-39        41
40+          13

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [52]:
purchase_data["Age Ranges"] = pd.cut(purchase_data["Age"], age_bin, labels=age_labels)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Ranges  
0  Extraction, Quickblade Of Trembling Hands   3.53      20-24  
1                          Frenzied Scimitar   1.56        40+  
2                               Final Critic   4.88      20-24  
3                                Blindscythe   3.27      20-24  
4                                       Fury   1.44      20-24

In [53]:
#GroupBy age range
grouped_age_range = purchase_data.groupby(["Age Ranges"])

In [54]:
#Calculate purchase count by age range
gen_purchase_count = grouped_age_range["Purchase ID"].count()
gen_purchase_count


Age Ranges
<10       23
10-14     28
15-19    136
20-24    365
25-29    101
30-34     73
35-39     41
40+       13
Name: Purchase ID, dtype: int64

In [55]:
#Calculate average purchase price by age
age_avg_purchase_price = round(grouped_age_range["Price"].mean(),2)
age_avg_purchase_price


Age Ranges
<10      3.35
10-14    2.96
15-19    3.04
20-24    3.05
25-29    2.90
30-34    2.93
35-39    3.60
40+      2.94
Name: Price, dtype: float64

In [56]:
#Calculate average total purchase price per person by age range
age_avg_total_purchase = round(grouped_age_range.Price.sum(),2)
age_avg_total_purchase



Age Ranges
<10        77.13
10-14      82.78
15-19     412.89
20-24    1114.06
25-29     293.00
30-34     214.00
35-39     147.67
40+        38.24
Name: Price, dtype: float64

In [57]:
#Create a summary data frame to hold the results
purchasing_analysis_age_df = pd.DataFrame({"Purchase Count": gen_purchase_count, "Average Purchase Price per Person by Age Range": age_avg_purchase_price, "Age Range Total Purchase Price": age_avg_total_purchase})

purchasing_analysis_age_df["Average Purchase Price per Person by Age Range"] = purchasing_analysis_age_df["Average Purchase Price per Person by Age Range"].map("${:,.2f}".format)
purchasing_analysis_age_df["Age Range Total Purchase Price"] = purchasing_analysis_age_df["Age Range Total Purchase Price"].map("${:,.2f}".format)

#Display the summary data frame
purchasing_analysis_age_df

Purchase Count Average Purchase Price per Person by Age Range  \
Age Ranges                                                                  
<10                     23                                          $3.35   
10-14                   28                                          $2.96   
15-19                  136                                          $3.04   
20-24                  365                                          $3.05   
25-29                  101                                          $2.90   
30-34                   73                                          $2.93   
35-39                   41                                          $3.60   
40+                     13                                          $2.94   

           Age Range Total Purchase Price  
Age Ranges                                 
<10                                $77.13  
10-14                              $82.78  
15-19                             $412.89  
20-24                           $1,114.06  
25-29                             $293.00  
30-34                             $214.00  
35-39                             $147.67  
40+                                $38.24

## Top Spenders

* Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
  * SN
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  
* Run basic calculations to obtain the results in the table below
* Create a summary data frame to hold the results
* Sort the total purchase value column in descending order
* Optional: give the displayed data cleaner formatting
* Display a preview of the summary data frame



In [58]:
purchase_data.columns

Index(['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price',
       'Age Ranges'],
      dtype='object')

In [70]:
#GroupBy SN
grouped_spenders = purchase_data.groupby(["SN"])
grouped_spenders.head()

Purchase ID               SN  Age                 Gender  Item ID  \
0              0          Lisim78   20                   Male      108   
1              1      Lisovynya38   40                   Male      143   
2              2       Ithergue48   24                   Male       92   
3              3    Chamassasya86   24                   Male      100   
4              4        Iskosia90   23                   Male      131   
5              5          Yalae81   22                   Male       81   
6              6        Itheria73   36                   Male      169   
7              7      Iskjaskst81   20                   Male      162   
8              8        Undjask33   22                   Male       21   
9              9      Chanosian48   35  Other / Non-Disclosed      136   
10            10        Inguron55   23                   Male       95   
11            11     Haisrisuir60   23                   Male      162   
12            12     Saelaephos52   21                   Male      116   
13            13      Assjaskan73   22                   Male        4   
14            14      Saesrideu94   35                   Male      165   
15            15        Lisassa64   21                 Female       98   
16            16        Lisirra25   20                   Male       40   
17            17        Zontibe81   21                   Male      161   
18            18        Reunasu60   22                 Female       82   
19            19        Chamalo71   30                   Male       89   
20            20     Iathenudil29   20                   Male       57   
21            21    Phiarithdeu40   20                   Male      168   
22            22     Siarithria38   38  Other / Non-Disclosed       24   
23            23         Eyrian71   40                   Male      151   
24            24          Siala43   30                   Male      141   
25            25        Lisirra87   29                   Male      178   
26            26       Lirtossa84   11                   Male       71   
27            27          Eusri44    7                   Male       96   
28            28           Aela59   21                   Male      119   
29            29          Tyida79   24                   Male       37   
..           ...              ...  ...                    ...      ...   
750          750         Iljask75   22                   Male      167   
751          751      Lisjaskan36   11                   Male      180   
752          752    Mindjasksya61   17                   Male      112   
753          753   Frichirranya75   36                   Male      178   
754          754  Pheosurllorin41   23                 Female       79   
755          755         Raesty92   12                   Male       76   
756          756          Ilast79   20                   Male       73   
757          757        Eratiel90   18                   Male       57   
758          758           Iral74   21                   Male      182   
759          759      Tyaelorap29   25                   Male       72   
760          760       Aithelis62   21                   Male       44   
761          761          Assim27   45                   Male       17   
762          762           Asur53   26                   Male      110   
763          763     Lassilsala30   21                   Male       85   
764          764     Saedaiphos46   18                   Male      113   
765          765          Irith83   18                   Male      130   
766          766     Aelastirin39   23                   Male       58   
767          767          Ilmol66    8                 Female       92   
768          768     Assassasta79   38                   Male       92   
769          769        Ilosian36   15                   Male      145   
770          770       Lirtosia63   34                   Male       12   
771          771     Iskossasda43   16                   Male       25   
772      

In [68]:
#GroupBy SN
purchase_data.groupby(["SN"]).sum().head()


Purchase ID  Age  Item ID  Price
SN                                            
Adairialis76          467   16      123   2.28
Adastirin33           142   35      175   4.48
Aeda94                388   17      128   4.91
Aela59                 28   21      119   4.32
Aelaria33             630   23      171   1.79

In [61]:
# Calculate top spender
top_spender_total = grouped_spenders.sum()["Price"]
top_spender_total.sort_values(ascending=False).head()

SN
Lisosia93      18.96
Idastidru52    15.45
Chamjask73     13.83
Iral74         13.62
Iskadarya95    13.10
Name: Price, dtype: float64

In [62]:
# Calculate top spender averages
top_spender_average = grouped_spenders.mean()["Price"]
top_spender_average.head()


SN
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
Name: Price, dtype: float64

In [63]:
# Calculate top spender counts
top_spender_count = grouped_spenders.count()["Price"]
top_spender_count.sort_values(ascending=False).head()

SN
Lisosia93      5
Iral74         4
Idastidru52    4
Aina42         3
Iri67          3
Name: Price, dtype: int64

In [64]:
# Dataframe for top spender purchasing patterns
top_spender_data_df = pd.DataFrame({"Total Purchased Amount": top_spender_total,
                          "Average Purchase Price": top_spender_average,
                          "Purchase Count": top_spender_count})

# Display top 5 spenders' purchasing patterns df- sorted in descending order
top_spender_data_df.sort_values("Total Purchased Amount", ascending=False, inplace=True)

top_spender_data_df["Total Purchased Amount"] = top_spender_data_df["Total Purchased Amount"].map("${:,.2f}".format)
top_spender_data_df["Average Purchase Price"] = top_spender_data_df["Average Purchase Price"].map("${:,.2f}".format)

top_spender_data_df.head()

Total Purchased Amount Average Purchase Price  Purchase Count
SN                                                                       
Lisosia93                   $18.96                  $3.79               5
Idastidru52                 $15.45                  $3.86               4
Chamjask73                  $13.83                  $4.61               3
Iral74                      $13.62                  $3.40               4
Iskadarya95                 $13.10                  $4.37               3

## Most Popular Items

## Most Popular Items

* Identify the 5 most profitable items by total purchase value, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value
  
* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [65]:
purchase_data.columns

Index(['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price',
       'Age Ranges'],
      dtype='object')

In [66]:
# #Most popular items purchased
grouped_purchase_items = purchase_data.groupby(["Item ID", "Item Name"])

top_item_qty_sold = grouped_purchase_items.count()["Price"]
top_item_avgprice = grouped_purchase_items.mean()["Price"]
top_item_revenues = grouped_purchase_items.sum()["Price"]

# Dataframe for top spender purchasing patterns
top_items_data_df = pd.DataFrame({"Item Total Revenue": top_item_revenues, 
                          "Item Average Price": top_item_avgprice,
                          "Total Qty Sold": top_item_qty_sold})

# # Display top 5 spenders' purchasing patterns df- sorted in descending order
top_items_data_df.sort_values("Total Qty Sold", ascending=False)

top_items_data_df["Item Average Price"] = top_items_data_df["Item Average Price"].map("${:,.2f}".format)
top_items_data_df["Item Total Revenue"] = top_items_data_df["Item Total Revenue"].map("${:,.2f}".format)
top_items_data_df.head(5)



Item Total Revenue Item Average Price  \
Item ID Item Name                                                  
0       Splinter                        $5.12              $1.28   
1       Crucifer                        $9.78              $3.26   
2       Verdict                        $14.88              $2.48   
3       Phantomlight                   $14.94              $2.49   
4       Bloodlord's Fetish              $8.50              $1.70   

                            Total Qty Sold  
Item ID Item Name                           
0       Splinter                         4  
1       Crucifer                         3  
2       Verdict                          6  
3       Phantomlight                     6  
4       Bloodlord's Fetish               5

## Most Profitable Items

### Most Profitable Items

* Identify the 5 most profitable items by total purchase value, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value



In [67]:
# Item Table (Sorted by Total Purchase Value)
top_items_data_df.sort_values("Total Qty Sold", ascending=False).head(5)


Item Total Revenue  \
Item ID Item Name                                                         
178     Oathbreaker, Last Hope of the Breaking Storm             $50.76   
145     Fiery Glass Crusader                                     $41.22   
108     Extraction, Quickblade Of Trembling Hands                $31.77   
82      Nirvana                                                  $44.10   
19      Pursuit, Cudgel of Necromancy                             $8.16   

                                                     Item Average Price  \
Item ID Item Name                                                         
178     Oathbreaker, Last Hope of the Breaking Storm              $4.23   
145     Fiery Glass Crusader                                      $4.58   
108     Extraction, Quickblade Of Trembling Hands                 $3.53   
82      Nirvana                                                   $4.90   
19      Pursuit, Cudgel of Necromancy                             $1.02   

                                                      Total Qty Sold  
Item ID Item Name                                                     
178     Oathbreaker, Last Hope of the Breaking Storm              12  
145     Fiery Glass Crusader                                       9  
108     Extraction, Quickblade Of Trembling Hands                  9  
82      Nirvana                                                    9  
19      Pursuit, Cudgel of Necromancy                              8

Observations:
1) Marketing should focus on males between the age ranges of 15-29 to impact sales volumes for items averaging $3 which make up the the bulk of sales.

2) Special consideration for premium items should be given to purchases between the ages 35-39 who have the most purchasing power and spend the highest average cost per item at $3.60; there is likely a correlation between purchasers under the age of 10 and purchasers between the ages of 35-39 as the purchasers under 10 are the second highest group that had the highest average spent on individual items. 

An assumption could be made that purchasers under the age of 10 have parents between the ages of 35-39 and are willing pay more for individual items due to a greater source of income. As these purchasers mature into adolesence the price they are willing to spend per item declines and begins to average around $3.00.

3) Individual lower priced items are not the most purchased items; purchases are driven by another value other than price
